src/repr/src/network_policy_id.rs

```diff
@@ -0,0 +1,116 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::fmt;
+use std::str::FromStr;
+
+use anyhow::{anyhow, Error};
+use columnation::{Columnation, CopyRegion};
+use mz_lowertest::MzReflect;
+use proptest_derive::Arbitrary;
+use serde::{Deserialize, Serialize};
+
+const SYSTEM_CHAR: char = 's';
+const USER_CHAR: char = 'u';
+
+/// The identifier for a network policy.
+#[derive(
+    Arbitrary,
+    Clone,
+    Copy,
+    Debug,
+    Eq,
+    PartialEq,
+    Ord,
+    PartialOrd,
+    Hash,
+    Serialize,
+    Deserialize,
+    MzReflect,
+)]
+pub enum NetworkPolicyId {
+    System(u64),
+    User(u64),
+}
+
+impl NetworkPolicyId {
+    pub fn is_system(&self) -> bool {
+        matches!(self, Self::System(_))
+    }
+
+    pub fn is_user(&self) -> bool {
+        matches!(self, Self::User(_))
+    }
+
+    pub fn is_builtin(&self) -> bool {
+        self.is_system()
+    }
+}
+
+impl FromStr for NetworkPolicyId {
+    type Err = Error;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        fn parse_u64(s: &str) -> Result<u64, Error> {
+            if s.len() < 2 {
+                return Err(anyhow!("couldn't parse network policy id '{s}'"));
+            }
+            s[1..]
+                .parse()
+                .map_err(|_| anyhow!("couldn't parse network policy  id '{s}'"))
+        }
+
+        match s.chars().next() {
+            Some(SYSTEM_CHAR) => {
+                let val = parse_u64(s)?;
+                Ok(Self::System(val))
+            }
+            Some(USER_CHAR) => {
+                let val = parse_u64(s)?;
+                Ok(Self::User(val))
+            }
+            _ => Err(anyhow!("couldn't parse network policy  id '{s}'")),
+        }
+    }
+}
+
+impl fmt::Display for NetworkPolicyId {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        match self {
+            Self::System(id) => write!(f, "{SYSTEM_CHAR}{id}"),
+            Self::User(id) => write!(f, "{USER_CHAR}{id}"),
+        }
+    }
+}
+
+impl Columnation for NetworkPolicyId {
+    type InnerRegion = CopyRegion<NetworkPolicyId>;
+}
+
+#[mz_ore::test]
+fn test_network_poliyc_id_parsing() {
```

Comment:
```suggestion
fn test_network_policy_id_parsing() {
```

Reply:
Not blocking but an easy thing to do is add a `proptest!` that generates arbitrary `NetworkPolicyId`s and asserts they roundtrip through the string impls